[API] Issues/issue 529 driverscan additional structures

fgomulka opened the pull request:
N.B. This PR requires a major version change which is not currently part of the PR.

This patches an issue for some versions of windows that require the bottom-up approach for finding the size of "object body" by recognizing additional structures occupying the "object body" position. The sizes of these additional structures are added to the size of `_DRIVER_OBJECT` to more accurately compute the actual size of "object body".

This PR resolves #529

ikelos committed:
Core: Support zip files as plugin paths

ikelos committed:
Core: Change zipimporting code

The previous patch in this branch would allow passing zipfiles
as part of the plugin path list.  Instead, we now traverse any
zipfiles in the plugin directories and load them based on where
they are.  The filenames for the zip files aren't important.

atcuno committed:
Initial commit for mass testing

Commit:
Catch exceptions triggered during testing

Commit:
Fix typos and add more debug statements

atcuno committed:
Commit all requested changes except those related to array handling/creating

atcuno committed:
Updates and bug fixes

ikelos committed:
Windows: Fix up typo in num_entries parameter

ikelos committed:
Core: Add in offline constant

ikelos committed:
Automagic: Add in remote banner cache first attempt

ikelos committed:
Automagic: Support remote banner versioning and chaining

ikelos committed:
Automagic: Catch remote download exceptions

fgomulka committed:
Add patch for driverscan for windows 7 and earlier

This patches an issue for some versions of windows that require the
bottom-up approach for finding the size of "object body" by
recognizing additional structures occupying the "object body"
position. The sizes of these additional structures are added to
the size of `_DRIVER_OBJECT` to more accurately compute the actual
size of "object body".

fgomulka committed:
Add patch for driverscan for windows 7 and earlier

This patches an issue for some versions of windows that require the
bottom-up approach for finding the size of "object body" by
recognizing additional structures occupying the "object body"
position. The sizes of these additional structures are added to
the size of `_DRIVER_OBJECT` to more accurately compute the actual
size of "object body".

fgomulka committed:
Comment and Pydoc changes

fgomulka committed:
Comment and Pydoc changes

ikelos committed:
Core: Add module collection support

ikelos committed:
Automagic: Make symbol_shift automagic change

ikelos committed:
Layers: Add minimum framework requirements

ikelos committed:
Linux: Update all plugins to ModuleRequirement

ikelos committed:
Mac: Update all plugins to ModuleRequirement

ikelos committed:
Windows: Convert pslist as an example

ikelos committed:
Windows: Refactor PDB code and support VS19 new types

ikelos committed:
Windows: Ensure we emit LF_STRUCTURE_VS19 types

fgomulka commented:
@ikelos I've noticed these merge conflicts have been here a while, and I've tried to resolve them on my own CL. Do you have any recommendations for resolving these myself?

I think the issue has to do with the fact that `volatility3/framework/plugins/windows/poolscanner.py` has been moved to `volatility3/plugins/windows/poolscanner.py` so that `pool.py` can use its type `PoolConstraint`

ikelos commented:
Hiya @fgomulka, the poolscanner.py file still lives in the same location it always has (`volatility3/framework/plugins/windows/poolscanner.py`), I think the changes causing a conflict are probably the sweeping python3.6 changes including format strings and the typing annotation support in 3.6 onwards.  I think you'll be able to rebase it without problem.  Make sure that your version of `develop` is in check with the main volatilityfoundation one.  Then you can checkout your local branch and do `git rebase develop`.  You'll need to use `git megetool` after that and make sure that the final version looks as you'd expect (containing both your and our changes) and then do `git rebase --continue`.  You'll need to do this until there's no more issues.  Then once you're happy with the final commit, you can force push it back up here with `git push -f`.  Please give that a try and let me know if there's any problems...  5:)

atcuno committed:
Fix several code paths that could infinite loop and were triggered based on @fgomulka testing

fgomulka commented:
Thanks for the help! Looks like I wasn't rebasing from a branch even with  `volatilityfoundation:develop`. Looks like that solved the conflicts.

fgomulka commented:
It seems that the fix for driverscan on xp has broken. I do not believe this issue is related to the code added to this branch, but rather everything that got merged into this branch. Any idea what might have gone wrong? Or is this issue not reproducible?

ikelos committed:
Windows: Fix pdbconv handling of unknown types

ikelos committed:
Windows: Update netscan offsets for win10-17763-x64 (#478)

ikelos reviewed (commented):
Cool, looks good, but since it is an API change, it'll need to wait for the major version number to bump...

ikelos commented:
I saw a comment in my email about an issue with xp, was that resolved or is it something we need to consider?

fgomulka commented:
The XP issue was resolved. Everything works great from my side of things for XP.

ikelos committed:
Windows: Ensure the kernel contains an actual kvo

ikelos committed:
Windows: Pdbsan typing and doc fixes

ikelos committed:
Merge pull request #540 from volatilityfoundation/feature/pdbconv-vs19

Feature/pdbconv vs19

ikelos committed:
Windows: Mute crypto warnings from lgtm

ikelos committed:
Windows: Fix memmap not to create empty files

ikelos committed:
Windows: Add coalesce support to memmap

ikelos committed:
Core: Include API changes file

ikelos committed:
Core: Don't get to hasty with the API version

ikelos committed:
Merge pull request #538 from volatilityfoundation/feature/better-modules-support2

Feature/better modules support

ikelos committed:
Linux: Support None to turn off remote banner locations

ikelos committed:
Automagic: Minor fixes for jar files and layers

ikelos committed:
Layers: Coalesce intel mapping responses

atcuno committed:
Fix return types, change confusing use of continue, switch to get_absolute_symbol_address

ikelos committed:
Merge pull request #484 from volatilityfoundation/skeleton_key_check_plugin

Skeleton key check plugin

atcuno committed:
Remove unused variable from plugin

atcuno committed:
Add missing patch that fixes KERB_ECRYPT structure size

ikelos committed:
Core: Ensure module configuration is recordable

ikelos committed:
Windows: Update to kernel module for ease of config

ikelos committed:
Windows: Remove hardcoded parameter in poolscanner

ikelos committed:
Core: Fix caching of pointer objects

ikelos committed:
Generic: YaraScanner fail on no rules

ikelos committed:
Windows: Fix pstree physical/virtual offsets

ikelos committed:
Merge pull request #551 from volatilityfoundation/feature/windows-kernel-module

Windows: Update to kernel ModuleRequirement for ease of config

ikelos committed:
Plugins: Add in architectures for ModuleRequirement

Fixes #553

ikelos committed:
Automagic: Only traverse down intel layers

ikelos committed:
Linux: Shift linux plugins to using kernel as their default module

ikelos committed:
Mac: Standadize on kernel config name

ikelos committed:
Merge pull request #549 from volatilityfoundation/feature/intel-coalesce-by-default

Layers: Coalesce intel mapping responses

ikelos committed:
Core: Make pointer caching more accurate

ikelos committed:
Windows: Ensure DTBs don't have all valid entries

We were finding that PAE results were being masked by certain pages with
64bit results.  The DTBs could be distinguished from good DTBs because
every entry was valid (which for a 64-bit or even 32-bit DTB is
extremely unlikely except on systems with *vast* quanties of memory.

As such, this now tests for all entries being allocated and doesn't return the
DTB if this is the case.

ikelos committed:
Automagic: Remove the superfluous num_entries fields

ikelos committed:
Objects: Fix up pointer caching to default to native layer

ikelos committed:
Merge pull request #554 from volatilityfoundation/feature/linux-mac-windows-unify-config

Feature/linux mac windows unify config

ikelos committed:
CLI: Verify improvement for buildbot

ikelos committed:
Revert "CLI: Verify improvement for buildbot"

This fix for buildbot didn't seem to have any effect

This reverts commit 57e1243932168b7c3c179ac4aba7507d8af483dd.

ikelos committed:
Merge pull request #526 from volatilityfoundation/feature/linux-mac-repo

Feature/linux mac repo

shu-tom committed:
Fixed an issue where the PDB guid value was not zero padded.

ikelos committed:
Merge pull request #555 from shu-tom/develop

Fixed an issue where the symbol file download from the Microsoft site failed

ikelos committed:
Windows: Pstree loop protection

ikelos committed:
Merge pull request #511 from volatilityfoundation/issues/issue505-2

Windows: Ensure DTBs don't have all valid entries

ikelos committed:
Windows: Reduce the full automagic range *again*  5:\

ikelos committed:
Windows: Totally revert broken merge issue505-2

ikelos committed:
Automagic: Clean up the windows self-referential dtb finder

ikelos committed:
Automagic: Add a section for where old kernels live

ikelos committed:
Automagic: Increase windows self-ref segment for win7

ikelos committed:
Automagic: Remove unused old code

ikelos committed:
Automagic: Further sections for older windows

ikelos committed:
Automagic: Better check for bad DTBs

superponible committed:
correct usage for RegValueTypes enum

ikelos committed:
Linux: Fixes after the pslist kernel req change

ikelos committed:
Windows: Fix issue #385

ikelos committed:
Merge pull request #561 from volatilityfoundation/issues/issue_547_regvaluetype_usage

Correct usage for RegValueTypes enum

ikelos committed:
Merge pull request #557 from volatilityfoundation/issues/issue556

Windows: Pstree loop protection

ikelos committed:
Linux: Remove unused code (LGTM)

ikelos committed:
Mac: Remove unused code (LGTM)

ikelos committed:
Merge pull request #406 from volatilityfoundation/plugin-zips

Plugin zips

ikelos committed:
Mac: Finish the previous LGTM clean-up

ikelos committed:
Automagic: Extended the possible DTB locations for Win 11

Commit:
Doc: list_head.to_list, pslist pid 0

Added Documentation for the params of list_head.to_list

Added a comment to linux.pslist explaining why the init task is not yielded.

ikelos committed:
Automagic: Improve reserved bit detection for PAE

ikelos committed:
Automagic: Sort DTB results by test

ikelos committed:
Revert "Automagic: Sort DTB results by test"

This reverts commit 75feba009918906f99b5a5cc294e56442b6f263d.

ikelos committed:
Automagic: Actually use reserved_bits

ikelos committed:
Automagic: Windows 32-bit bit 7 is not always 0

ikelos committed:
Automagic: Prioritize tests in order

ikelos committed:
Automagic: Ensure reserved bits are for valid pages

ikelos committed:
Automagic: Don't churn memory as much

ikelos committed:
Merge pull request #563 from NiklasBeierl/doc_list_head

Doc: list_head.to_list, pslist pid 0

ikelos committed:
Merge pull request #559 from volatilityfoundation/feature/unify-dtb-selfref-checks

Feature/unify dtb selfref checks

ikelos committed:
Merge pull request #546 from volatilityfoundation/mac_list_files_infinite_loops

Fix several code paths that could infinite loop and were triggered ba…

ikelos committed:
Core: Fix module size finding

ikelos committed:
Linux: Fix kmsg plugin

ikelos commented:
Could we please get this rebased in preparation for merging?

fgomulka committed:
Merge branch 'issues/issue_529_driverscan_additional_structures' of github.com:fgomulka/volatility3 into issues/issue_529_driverscan_additional_structures

fgomulka commented:
Rebased!

ikelos commented:
That seems to have changed an awful lot of files now?  I think this may not have rebased properly?

fgomulka commented:
Hm, you're right, I'm not sure why all those past 100-ish commits showed up (and why they show a diff from the `volatilityfoundation:develop` branch). Is there a way to squash all the commits that are showing up from my rebase? 

Here's the sequence of steps I took--could you please let me know if I did something out of the ordinary?

1. Fetched/merged `volatilityfoundation:develop` into `fgomulka:develop`.
2. pulled these changes locally, then rebased this branch from `fgomulka:develop`
3. Manually solved 1 merge issue
4. Completed merge and pushed the changes.

ikelos commented:
Hmmm, I'm not certain, @npetroni is our git guru.  It looks like it's checking changes from an old `volatilityfoundation:develop` commit and the latest `fgomulka:/issues/issue_529_driverscan_additional_structures` branch?  I guess you could try rebasing it directly against `volatilityfoundation:develop`?

ikelos commented:
Either that, or you need to push your local `fgomulka:develop` as well as `fgomulka:/issues/issue_529_driverscan_additional_structures` back up to github?

fgomulka commented:
Yeah, both local branch are already pushed/up to date with the remote. I will try rebasing with `volatilityfoundation:develop`.

fgomulka commented:
It looks like I don't have the correct access rights to rebase directly from `volatilityfoundation:develop`. Is there a way that I can do this without fetching?

ikelos commented:
It should be possible, but I don't know the details.  5:S  At this point it may just be easier opening a new pull request and closing this one off?  We can still refer back to it and I think all the comments in this one had been addressed anyway?

fgomulka commented:
sounds good. I can open a new PR.

ikelos commented:
Thanks, sorry for all the fuss...  5:S
